File: leetcode-28-Find_the_Index_of_the_First_Occurrence_in_a_String/leetcode-28-Find_the_Index_of_the_First_Occurrence_in_a_String.cpp
#include <iostream>
#include <cstdlib>
#include <vector>
#include <string>


int strStr(std::string haystack, std::string needle) {
    std::vector<int> sub_vals(needle.size(),0);

    for(size_t i=1, j=0; i<sub_vals.size() && j<sub_vals.size();)
    {
        if(needle[i] == needle[j])
        {
            sub_vals[i++] = ++j;
        } else {
            if(j)
            {
                j = sub_vals[j-1];
            } else {
                ++i;
            }
        }
    }

    size_t i=0,
           j=0;
    while(j<haystack.size())
    {
        if(needle[i] == haystack[j])
        {
            ++i;
            ++j;
        }
        if(i==needle.size())
        {
            return j-i;
        }
        if (j < haystack.size() && needle[i] != haystack[j])
        {
            if(i)
            {
                i = sub_vals[i-1];
            } else {
                ++j;
            }
        }
    }
    return -1;
}


int main(){
    std::string haystack = "sadbutsad",
                needle = "sad";

    auto res = strStr(haystack, needle);
    
    std::cout << res << std::endl;
    
    return EXIT_SUCCESS;
}

File: leetcode-28-Find_the_Index_of_the_First_Occurrence_in_a_String/leetcode-28-Find_the_Index_of_the_First_Occurrence_in_a_String.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        pattern_lst: list[int] = [0] * len(needle)
        i_begin: int = 0
        for i in range(1, len(needle)):
            while needle[i_begin] != needle[i] and i_begin != 0:
                i_begin = pattern_lst[i_begin-1]
            else:
                if needle[i_begin] != needle[i]:
                    continue
            i_begin += 1
            pattern_lst[i] = i_begin
        
        i_haystack: int = 0
        i_needle: int = 0
        while i_needle < len(needle) and i_haystack < len(haystack):
            if needle[i_needle] == haystack[i_haystack]:
                i_haystack += 1
                i_needle += 1
            else:
                while needle[i_needle] != haystack[i_haystack] and i_needle != 0:
                    i_needle = pattern_lst[i_needle - 1]
                else:
                    if needle[i_needle] != haystack[i_haystack]:
                        i_haystack += 1
        if i_needle == len(needle):
            return i_haystack - len(needle)
        else:
            return -1


if __name__ == "__main__":
    haystack: str = "sadbutsad"
    needle: str = "sad"
    result = Solution().strStr(haystack, needle)
    print(result)
